### PR TITLE
SRL defaults and initializers

### PR DESCRIPTION
**Pull request by DeNeutoy**
- Add default hyperparams for SRL
- Add SRL Initializers
- Fix issue with sorting `Variables`
- Add a `experiment_configs` directory and param file for the SRL model
- Remove `batch_size` from trainer as it should be in `DataIterators`.
- Add sequence length masks to all encoders in all models.
- Add grad norm to `Trainer`.

### CODE REVIEW

File: allennlp/common/tensor.py [hunk 1]
```diff
@@ -44,6 +44,11 @@ def sort_batch_by_length(tensor: torch.FloatTensor, sequence_lengths: torch.Long
     restoration_indices : torch.LongTensor
         Indices into the sorted_tensor such that ``sorted_tensor[restoration_indices] == original_tensor``
     """
+    # Pytorch can't handle sorting with Variables.
+    # TODO(Mark): Assert whether this is allowed - does it break autograd?
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: Assert?  You mean figure out?

And yeah, I'm assuming this breaks autograd.  And even if we _can_ do it this way, wouldn't we need to wrap this in a `Variable` when we're done?

File: allennlp/models/semantic_role_labeler.py [hunk 2]
```diff
@@ -51,6 +55,8 @@ def __init__(self, vocab: Vocabulary,
         self.stacked_encoder = stacked_encoder
         self.tag_projection_layer = TimeDistributed(Linear(self.stacked_encoder.get_output_dim(),
                                                            self.num_classes))
+        initializer(self.stacked_encoder)
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: Would calling `initializer(self)` accomplish the same thing here?

File: allennlp/models/semantic_role_labeler.py [hunk 3]
```diff
@@ -191,8 +198,43 @@ def get_viterbi_pairwise_potentials(self):
 
     @classmethod
     def from_params(cls, vocab: Vocabulary, params: Params) -> 'SemanticRoleLabeler':
-        text_field_embedder = TextFieldEmbedder.from_params(vocab, params.pop("text_field_embedder"))
-        stacked_encoder = Seq2SeqEncoder.from_params(params.pop("stacked_encoder"))
+        """
+        With an empty ``params`` argument, this will instantiate a SRL model with the same
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: `s/a SRL/an SRL/`?

File: allennlp/models/semantic_role_labeler.py [hunk 4]
```diff
@@ -191,8 +198,43 @@ def get_viterbi_pairwise_potentials(self):
...
+        configuration as published in the Deep Semantic Role Labeling - What works and what's
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: Quotes around the title?

File: allennlp/models/semantic_role_labeler.py [hunk 5]
```diff
@@ -191,8 +198,43 @@ def get_viterbi_pairwise_potentials(self):
...
+        next  paper, as long as you've set ``allennlp.common.constants.GLOVE_PATH`` to the
+        location of your gzipped 100-dimensional glove vectors.
+
+        If you want to change parameters, the keys in the ``params`` object must match the
+        constructor arguments above.
+        """
+        default_embedder_params = {
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: I'm thinking there's got to be a better way to do this.  We're going to want to have configuration files that specify training datasets and model parameters for these models, and we'll be duplicating ourselves here and there.  We'll definitely want the full model parameters specified there, to make the connection to the data loading obvious, and just for ease of someone browsing through the configuration files.

What if we allow `params` to be `None` here, and if it's `None`, we load the parameters from that file?  This would be changed later, once we actually have such files.

File: allennlp/training/trainer.py [hunk 6]
```diff
@@ -57,6 +56,8 @@ def __init__(self,
             An integer specifying the CUDA device to use. If -1, the CPU is used.
             Multi-gpu training is not currently supported, but will be once the
             Pytorch DataParallel API stabilises.
+        grad_norm: float, optional, (default = None).
+            If provided, gradients will be rescaled to have a maximum of this value.
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: `gradient norms will be rescaled`?

File: experiment_config/semantic_role_labeler.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    "type": "adadelta",
+    "lr": 0.000001,
+    "rho": 0.95
+  },
+  "num_epochs": 500,
+  "grad_norm": 1.0,
+  "patience" : 500,
+  "serialization_prefix" : "/path/to/save/directory/replace me!",
+  "cuda_device" : 0
+}
```
Comment (matt-gardner) [inline experiment_config/semantic_role_labeler.json]: File should have trailing newline.

File: allennlp/models/decomposable_attention.py [hunk 8]
```diff
@@ -116,11 +116,13 @@ def forward(self,  # type: ignore
...
         hypothesis_mask = get_text_field_mask(hypothesis).float()
+        premise_sequence_lengths = get_lengths_from_binary_sequence_mask(premise_mask)
+        hypothesis_sequence_lengths = get_lengths_from_binary_sequence_mask(hypothesis_mask)
 
         if self._premise_encoder:
-            embedded_premise = self._premise_encoder(embedded_premise)
+            embedded_premise = self._premise_encoder(embedded_premise, premise_sequence_lengths)
         if self._hypothesis_encoder:
-            embedded_hypothesis = self._hypothesis_encoder(embedded_hypothesis)
+            embedded_hypothesis = self._hypothesis_encoder(embedded_hypothesis, hypothesis_sequence_lengths)
```
Comment (matt-gardner) [inline allennlp/models/decomposable_attention.py]: We should override `forward` in `Seq2*Encoder` to specify what the expected API is.

Comment (matt-gardner) [inline allennlp/models/decomposable_attention.py]: Also, I think there's at least one encoder that was missed - the `TokenCharactersEmbedder` has an encoder that needs to be updated to use this.

Comment (matt-gardner) [inline allennlp/models/decomposable_attention.py]: That means the `CNNEncoder` needs to be updated to take the sequence lengths, also, to match the API.  We can just change `forward` to match, and ignore the argument with a `TODO` to handle it later.

File: allennlp/models/semantic_role_labeler.py (hunk 2 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeler.py]: We need to be careful doing that, because it will also initialize our carefully pre-initialized embedding layers. Given this model is simple, I didn't use the regex functionality of the `InitializerApplicator`, but I can do if you would prefer.

Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: The trouble is if someone passes in a `TextEmbedder` that uses a character-level encoder - the encoder won't get initialized.  It'd be nice to have some way for modules to say "I've already been initialized", so that the embeddings don't get messed up, but everything else works.

File: allennlp/models/semantic_role_labeler.py (hunk 5 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeler.py]: Something like that would be good. We also need to consider doing something like they have in `torchvision` where a pretrained model is autodownloaded and saved somewhere if you try to use a model, that would be nice.

File: allennlp/models/semantic_role_labeler.py (hunk 3 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeler.py]: Woah. This is super weird. You would say "an SRL model", but "a Semantic Role Labeling model".  Deep.

Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: https://english.stackexchange.com/questions/1016/do-you-use-a-or-an-before-acronyms

File: allennlp/common/tensor.py [hunk 9]
```diff
@@ -23,39 +25,39 @@ def get_lengths_from_binary_sequence_mask(mask: torch.ByteTensor):
...
-    # TODO(Mark): Assert whether this is allowed - does it break autograd?
-    if isinstance(sequence_lengths, Variable):
-        sequence_lengths = sequence_lengths.data
+
+    if not isinstance(tensor, Variable) or not isinstance(sequence_lengths, Variable):
+        raise ConfigurationError("Both the tensor and sequence lengths must be torch.autograd.Variables.")
 
     sorted_sequence_lengths, permutation_index = sequence_lengths.sort(0, descending=True)
-    sorted_tensor = tensor[permutation_index]
+    sorted_tensor = tensor.index_select(0, permutation_index)
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/common/tensor.py]: @DeNeutoy, I was curious if this worked, and so I checked.  Changing this back to `tensor[permutation_index]` still passes tests.  I ran a quick test with `timeit` to see which was faster:

```python
>>> import timeit
>>> timeit.timeit(setup='import torch; a = torch.rand(64, 14, 6); l = torch.rand(64); s, b = l.sort(0, descending=True)', stmt='t = a[b]', number=1000000)
2.71397556500051
>>> timeit.timeit(setup='import torch; a = torch.rand(64, 14, 6); l = torch.rand(64); s, b = l.sort(0, descending=True)', stmt='t = a.index_select(0, b)', number=1000000)
2.9591599720006343
```
Looks like it's about 10% faster to use advanced indexing instead of `index_select`.

Actually, scratch that.  I tried the same thing on the GPU (adding a call to `.cuda()` after the calls to `torch.rand()` in the setup), and the times were basically identical.

So the difference is just readability?  Any other reason to switch, @nelson-liu?

[the remark this answers is not in the document]
Comment (nelson-liu) [reply allennlp/common/tensor.py]: Nope, I was just curious (since I would have used advanced indexing). I forgot to follow up, but I believe the reason Mark used `index_select` is because you can't pass a `Variable` directly as advanced indexing input.

[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/common/tensor.py]: That may have been the issue back when this was merged, but it works now.  We could switch if we wanted to.  That's good to know for future reference.

[the remark this answers is not in the document]
Comment (DeNeutoy) [reply allennlp/common/tensor.py]: Sorry I only just saw all this! Yes the variable issue was originally why.
